agent/proxycfg/manager.go

```diff
@@ -0,0 +1,350 @@
+// certificates, intentions, and service discovery results for any declared
+// upstreams. See package docs for more detail.
+type Manager struct {
+	ManagerConfig
+
+	// stateCh is notified for any service changes in local state. We only use
+	// this to trigger on _new_ service addition since it has no data and we don't
+	// want to maintain a full copy of the state in order to diff and figure out
+	// what changed. Luckily each service has it's own WatchCh so we can figure
+	// out changes and removals with those efficiently.
+	stateCh chan struct{}
+
+	mu       sync.Mutex
+	started  bool
+	proxies  map[string]*state
+	watchers map[string]map[uint64]chan *ConfigSnapshot
+}
+
+// ManagerConfig holds the required external dependencies for a Manager
+// instance. All fields must be set to something valid or the manager will
+// panic. The ManagerConfig is passed by value to NewManager so the passed value
+// can be mutated safely.
+type ManagerConfig struct {
+	// Cache is the agent's cache instance that can be used to retrieve, store and
+	// monitor state for the proxies.
+	Cache *cache.Cache
+	// state is the agent's local state to be watched for new proxy registrations.
+	State *local.State
+	// source describes the current agent's identity, it's used directly for
+	// prepared query discovery but also indirectly as a way to pass current
+	// Datacenter name into other request types that need it. This is sufficient
+	// for now and cleaner than passing the entire RuntimeConfig.
+	Source *structs.QuerySource
+	// logger is the agent's logger to be used for logging logs.
+	Logger *log.Logger
+}
+
+// NewManager constructs a manager from the provided agent cache.
+func NewManager(cfg ManagerConfig) (*Manager, error) {
+	if cfg.Cache == nil || cfg.State == nil || cfg.Source == nil ||
+		cfg.Logger == nil {
+		return nil, errors.New("all ManagerConfig fields must be provided")
+	}
+	m := &Manager{
+		ManagerConfig: cfg,
+		// Single item buffer is enough since there is no data transferred so this
+		// is "level triggering" and we can't miss actual data.
+		stateCh:  make(chan struct{}, 1),
+		proxies:  make(map[string]*state),
+		watchers: make(map[string]map[uint64]chan *ConfigSnapshot),
+	}
+	return m, nil
+}
+
+// Run is the long-running method that handles state syncing. It should be run
+// in it's own goroutine and will continue until a fatal error is hit or Close
+// is called. Run will return an error if it is called more than once, or called
+// after Close.
+func (m *Manager) Run() error {
+	m.mu.Lock()
+	alreadyStarted := m.started
+	m.started = true
+	stateCh := m.stateCh
+	m.mu.Unlock()
+
+	// Protect against multiple Run calls.
+	if alreadyStarted {
+		return ErrStarted
+	}
+
+	// Protect against being run after Close.
+	if stateCh == nil {
+		return ErrStopped
+	}
+
+	// Register for notifications about state changes
+	m.State.Notify(stateCh)
+	defer m.State.StopNotify(stateCh)
+
+	for {
+		m.mu.Lock()
```

Comment:
Mostly an observation:

If I find myself doing something like:
```
func DoSomething() {
  for {
    mu.Lock()
    // lots of stuff
    mu.Unlock()
  }
}
```
I'll feel safer about avoiding possible lock/unlock problems in the face of stuff like `continue`, `return`, or `panic` by refactoring as:
```
func DoSomething() {
  for {
    doSomethingOnce()
  }
}
func doSomethingOnce() {
  mu.Lock()
  defer mu.Unlock()
  // lots of stuff
}
```

Reply:
Yeah I get nervy with locks in loops too. I judged this to be OK to start with as there were only a couple of lines between lock and unlock but that inevitably grew and now it's not immediately obvious that the loop between can't exit or continue the outer loop... I'll refactor that into a separate method for clarity